### PR TITLE
Complete TLS exchange improvements

### PR DESCRIPTION
**Pull request by yontyon**
In this PR we complete the improvements for TLS exchange procedure

- remove the restriction for no more than f exchanges at once
- allow master key based signed certificates in the client side
- refactor and improve the basic CRE mechanism
- refactor and fix the Apollo tests

### CODE REVIEW

File: bftclient/include/bftclient/StateControl.hpp [hunk 1]
```diff
@@ -0,0 +1,30 @@
+// Concord
+//
+// Copyright (c) 2020 VMware, Inc. All Rights Reserved.
```
Comment (nkumar04) [inline bftclient/include/bftclient/StateControl.hpp:3]: nit: 2022

File: bftclient/include/bftclient/config.h [hunk 2]
```diff
@@ -68,6 +68,7 @@ struct ClientConfig {
   RetryTimeoutConfig retry_timeout_config;
   std::optional<std::string> transaction_signing_private_key_file_path = std::nullopt;
   std::optional<concord::secretsmanager::SecretData> secrets_manager_config = std::nullopt;
+  std::optional<std::string> replicas_master_key_folder_path = "./replicas_rsa_keys";
```
Comment (nkumar04) [inline bftclient/include/bftclient/config.h:71]: nit: Should we name path as: "./replicas_keys" or "./replicas_master_keys"

File: bftclient/src/bft_client.cpp [hunk 3]
```diff
@@ -57,6 +58,13 @@ Client::Client(std::unique_ptr<bft::communication::ICommunication> comm, const C
   }
   communication_->setReceiver(config_.id.val, &receiver_);
   communication_->start();
+  if (config_.replicas_master_key_folder_path.has_value()) {
+    bft::communication::StateControl::instance().setGetPeerPubKeyMethod([&](uint32_t rid) {
+      concord::secretsmanager::SecretsManagerPlain psm_;
+      std::string key_path = config_.replicas_master_key_folder_path.value() + "/" + std::to_string(rid) + "/pub_key";
```
Comment (nkumar04) [inline bftclient/src/bft_client.cpp:64]: Do we exchange replica main public key if a replica is down in the initial start.
Do we need to add replica main key exchange here?

**Review by nkumar04 — APPROVED** [review]
LGTM!
Thanks!

File: bftclient/include/bftclient/config.h (hunk 2 above)
Comment (yontyon) [reply bftclient/include/bftclient/config.h:71]: we should, but as long as we work with rsa keys I prefer to specify this (to ease debugging and so)

File: bftclient/src/bft_client.cpp (hunk 3 above)
Comment (yontyon) [reply bftclient/src/bft_client.cpp:64]: right, thanks

**Pull request merged**